src/shared/Core/Authentication/MicrosoftAuthentication.cs

```diff
@@ -223,12 +224,25 @@ private async Task<AuthenticationResult> GetAccessTokenSilentlyAsync(IPublicClie
                 appBuilder.WithLogging(OnMsalLogMessage, LogLevel.Verbose, enablePiiLogging, false);
             }
 
-            // If we have a parent window ID we should tell MSAL about it so it can parent any authentication dialogs
-            // correctly. We only support this on Windows right now as MSAL only supports embedded/dialogs on Windows.
-            if (PlatformUtils.IsWindows() && !string.IsNullOrWhiteSpace(Context.Settings.ParentWindowId) &&
-                int.TryParse(Context.Settings.ParentWindowId, out int hWndInt) && hWndInt > 0)
+            // On Windows we should set the parent window handle for the authentication dialogs
+            // so that they are displayed as a child of the correct window.
+            if (PlatformUtils.IsWindows())
             {
-                appBuilder.WithParentActivityOrWindow(() => new IntPtr(hWndInt));
+                // If we have a parent window ID then use that, otherwise use the hosting terminal window.
+                IntPtr parentHandle;
+                if (!string.IsNullOrWhiteSpace(Context.Settings.ParentWindowId) &&
+                    int.TryParse(Context.Settings.ParentWindowId, out int hWndInt) && hWndInt > 0)
+                {
+                    parentHandle = new IntPtr(hWndInt);
+                }
+                else
+                {
+                    IntPtr consoleHandle = Kernel32.GetConsoleWindow();
+                    parentHandle = User32.GetAncestor(consoleHandle, GetAncestorFlags.GetRootOwner);
```

Comment:
Do we need NULL checks here?

Reply:
`GetConsoleWindow` will [return `NULL`](https://learn.microsoft.com/en-us/windows/console/getconsolewindow#return-value) (`IntPtr.Zero`) if there is no console window available.

The [docs for `GetAncestor`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getancestor) don't actually say what happens if you pass `NULL` as the `hwnd` argument. 🤔

However, trying it with `NULL` I get `NULL` back again (`IntPtr.Zero`). Using `IntPtr.Zero` with `WithParentActivityOrWindow` in MSAL results in an error since a window is absolutely required (for WAM).

Having no console window is totally possible, so I guess at this point all we can do is grab a handle to the desktop object?

cc: @bgavrilMS

Reply:
Added a new commit that, when the broker is enabled and we are unable to find a parent window, shows a 'dummy' top-level progress bar window, and uses that window handle as the parent.

<img width="801" alt="wam_dummy_window" src="https://user-images.githubusercontent.com/5658207/231294327-526aa279-daee-4f35-9336-a7ed13de4bc2.png">

Reply:
Ideally we should get this type of functionality into MSAL. Apps should need to reference Win.Forms and create windows. Will follow up offline.

Reply:
Note that this solution is imperfect, for example if multiple monitors are used.

Reply:
Another way of doing it is to fail yourself, i.e. if you are hosted in a UI application and the UI application does not initialize GCM correctly with its window handle - fail. If you are hosted in a console app, then the above logic should "just work"

Reply:
Since this scenario (no console window and not being told the parent window explicitly) is likely the default for most versions of Visual Studio and tools like SourceTree etc, we're going to need to support this out-of-the box.

I think implementing the 'dummy' progress dialog window should suffice.